<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
# Импорт необходимых библиотек
import pandas as pd
import psycopg2
from urllib.parse import quote_plus

In [2]:
# Параметры соединения заданы в файле config.txt
params = {k:v.strip() for k, v in (l.split('=') for l in open('config.txt'))}

In [3]:
connection = psycopg2.connect(**params)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [26]:
# Количество вакансий (vac_cnt)
query_3_1 = f'''
select 
    count(v.id) vac_cnt
from 
    public.vacancies v
'''

In [27]:
# результат запроса
df_01 = pd.read_sql_query(query_3_1, connection)
display(df_01)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\3432950440.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_01 = pd.read_sql_query(query_3_1, connection)


vac_cnt
0    49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [28]:
# Количество работодателей (emp_cnt)
query_3_2 = f'''
select 
    count(e.id) emp_cnt
from 
    public.employers e
'''

In [29]:
# результат запроса
df_02 = pd.read_sql_query(query_3_2, connection)
display(df_02)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\2320917537.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_02 = pd.read_sql_query(query_3_2, connection)


emp_cnt
0    23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [30]:
# Количество регионов (reg_cnt)
query_3_3 = f'''
select 
    count(a.id) reg_cnt
from 
    public.areas a
'''

In [31]:
# результат запроса
df_03 = pd.read_sql_query(query_3_3, connection)
display(df_03)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\227148523.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_03 = pd.read_sql_query(query_3_3, connection)


reg_cnt
0     1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [32]:
# Количество сфер деятельности (ind_cnt)
query_3_4 = f'''
select 
    count(i.id) ind_cnt
from 
    public.industries i
'''

In [33]:
# результат запроса
df_04 = pd.read_sql_query(query_3_4, connection)
display(df_04)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\3480038966.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_04 = pd.read_sql_query(query_3_4, connection)


ind_cnt
0      294

***

Выводы по предварительному анализу данных:

1. Всего в нашей базе примерно 49 тысяч вакансий и около 23,5 тысяч работодателей в 1362 регионах.
2. Количество сфер деятельности работодателей - 294.

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [34]:
# Количество вакансий (cnt) в каждом регионе (area)
query_4_1 = f'''
select 
    count(v.id) cnt,  
    a.name area
from 
    public.vacancies v     
join 
    public.areas a on v.area_id = a.id
group by 
    a.name
order by 
    1 desc
'''

In [35]:
# результат запроса
df_05 = pd.read_sql_query(query_4_1, connection)
display(df_05)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\1926707433.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_05 = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [36]:
# Количество вакансий с указанием минимальной или максимальной зарплаты (salary_cnt)
query_4_2 = f'''
select 
    count(v.id) salary_cnt
from 
    public.vacancies v   
where 
    salary_from is not null 
    or salary_to is not null
'''

In [37]:
# результат запроса
df_06 = pd.read_sql_query(query_4_2, connection)
display(df_06)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\1305001129.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_06 = pd.read_sql_query(query_4_2, connection)


salary_cnt
0       24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [38]:
# Средние значения для нижней (min_salary) и верхней границы (max_salary) зарплатной вилки
query_4_3 = f'''
select 
    round(avg(v.salary_from)) min_salary,
    round(avg(v.salary_to)) max_salary
from 
    public.vacancies v  
'''

In [39]:
# результат запроса
df_07 = pd.read_sql_query(query_4_3, connection)
display(df_07)

C:\Users\user\AppData\Local\Temp\ipykernel_15440\249990546.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_07 = pd.read_sql_query(query_4_3, connection)


min_salary  max_salary
0     71065.0    110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [6]:
# Количество вакансий (cnt) в зависимости от типа рабочего графика (schedule) и типа трудоустройства (employment)
query_4_4 = f'''
select 
    count(v.id) cnt,
    v.schedule,
    v.employment
from 
    public.vacancies v   
group by 
    v.schedule, v.employment
order by 
    1 desc
'''

In [7]:
# результат запроса
df_08 = pd.read_sql_query(query_4_4, connection)
display(df_08)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2982207268.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_08 = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [24]:
# Количество вакансий (cnt) в зависимости от требуемого опыта работы (exp)
query_4_5 = f'''
select 
    v.experience exp,
    count(v.id) cnt
from 
    public.vacancies v   
group by 
    v.experience
order by 2
'''

In [25]:
# результат запроса
df_09 = pd.read_sql_query(query_4_5, connection)
display(df_09)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2098640015.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_09 = pd.read_sql_query(query_4_5, connection)


exp    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Выводы по детальному анализу вакансий:

1. Количество вакансий зависит от региона.
Наибольшее количество вакансий в Москве и крупных городах-милионниках.
Примерно половина вакансий (24 из 49 тысяч) с указанием минимальной или максимальной зарплаты.

2. Среднее значение для нижней границы зарплатной вилки составляет около 71 тысячи рублей,
и среднее значение для верхней границы зарплатной вилки составляет более 110 тысяч рублей.

3. Количество вакансий зависит от типа рабочего графика и типа трудоустройства.
Наибольшее количество вакансий с рабочим графиком "Полный день" и типом  трудоустройства "Полная занятость", на втором месте - сочетание "Удаленная работа" и "Полная занятость".

4. Количество вакансий зависит от требуемого опыта работы.
Наибольшее количество вакансий требуют опыт "От 1 года до 3 лет", на втором месте - "От 3 до 6 лет", на третьем - "Нет опыта", на последнем - "Более 6 лет". Вероятно, что наиболее ценных сотрудников с большим опытом не ищут по объявлениям.


***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [39]:
# Топ 5 работодателей (emp_name) по количеству вакансий (cnt)
query_5_1 = f'''
select 
    e.name emp_name,
    count(v.id) cnt
from 
    public.vacancies v  
join 
    public.employers e on v.employer_id  = e.id
group by 1
order by 2 desc
limit 5
'''

In [40]:
# результат запроса
df_10 = pd.read_sql_query(query_5_1, connection)
display(df_10)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\190977165.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_10 = pd.read_sql_query(query_5_1, connection)


emp_name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
 Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [37]:
# Количество работодателей (emp_cnt) и вакансий (vac_cnt) для каждого региона (area)
query_5_2_0 = f'''
select 
	a.name area, 
	count(distinct e.id) emp_cnt, 
	count(distinct v.id) vac_cnt
from 
	public.areas a
left join public.employers e on e.area = a.id
left join public.vacancies v on v.area_id = a.id
group by a.id
order by 2 desc
'''

# Количество работодателей (emp_cnt) для регионов (area), в которых нет вакансий (vac_cnt)
query_5_2_1 = f'''
select 
    a.name area, 
    count(distinct e.id) emp_cnt, 
    count(distinct v.id) vac_cnt
from 
    public.areas a
left join 
    public.vacancies v on a.id = v.area_id
left join 
    public.employers e on a.id = e.area
where 
    v.area_id is null
group by 
    a.name
order by 2 desc
limit 1
'''

In [38]:
# результат запроса
df_11 = pd.read_sql_query(query_5_2_0, connection)
display(df_11)

# результат запроса
df_12 = pd.read_sql_query(query_5_2_1, connection)
display(df_12)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\1568678351.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_11 = pd.read_sql_query(query_5_2_0, connection)


area  emp_cnt  vac_cnt
0                                  Москва     5864     5333
1                         Санкт-Петербург     2217     2851
2                                   Минск     1115     2112
3                                  Алматы      721     1892
4                            Екатеринбург      609     1698
...                                   ...      ...      ...
1357                              Янино-1        0        0
1358                          Адыге-Хабль        0        1
1359                          Эркен-Шахар        0        0
1360  Индустриальный (Краснодарский край)        0        0
1361          Сабетта (Ямало-Ненецкий АО)        0        0

[1362 rows x 3 columns]

C:\Users\user\AppData\Local\Temp\ipykernel_16340\1568678351.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_12 = pd.read_sql_query(query_5_2_1, connection)


area  emp_cnt  vac_cnt
0  Россия      410        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [59]:
# Количество регионов (reg_cnt), в которых работодатель (emp_name) публикует свои вакансии
query_5_3 = f'''
select 
    e.name emp_name,
    count(distinct v.area_id) reg_cnt
from 
    public.employers e
left join 
    vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
'''

In [60]:
# результат запроса
df_13 = pd.read_sql_query(query_5_3, connection)
display(df_13)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\771739047.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_13 = pd.read_sql_query(query_5_3, connection)


emp_name  reg_cnt
0                      Яндекс      181
1                  Ростелеком      152
2                  Спецремонт      116
3      Поляков Денис Иванович       88
4                   ООО ЕФИН        71
...                       ...      ...
23170                  СДЕЛКА        0
23171            Alandr Group        0
23172                     СДК        0
23173      Lemon Land Lombard        0
23174                  Махина        0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [61]:
# Количество работодателей (emp_name), у которых не указана сфера деятельности (emp_ind_cnt)
query_5_4 = f'''
select 
    count(e.name) emp_name, 
    count(distinct ei.employer_id)  emp_ind_cnt
from 
    public.employers e
left join 
    employers_industries ei on e.id = ei.employer_id
where 
    ei.industry_id is null
'''

In [62]:
# результат запроса
df_14 = pd.read_sql_query(query_5_4, connection)
display(df_14)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2759350927.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_14 = pd.read_sql_query(query_5_4, connection)


emp_name  emp_ind_cnt
0      8419            0

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [63]:
# Количество работодателей (emp_name), у которых указано четыре сферы деятельности (ind_cnt)
query_5_5 = f'''
select 
    e.name emp_name, 
    count(distinct ei.industry_id) ind_cnt
from 
    public.employers e
left join 
    employers_industries ei on e.id = ei.employer_id
group by 
    e.name
having 
    count(distinct ei.industry_id) = 4
'''

In [64]:
# результат запроса
df_15 = pd.read_sql_query(query_5_5, connection)
display(df_15)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\3520188277.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_15 = pd.read_sql_query(query_5_5, connection)


emp_name  ind_cnt
0        101 Интернет        4
1            21vek.by        4
2                2ГИС        4
3                  2К        4
4         4 пикселя +        4
...               ...      ...
1127            ЮРИОН        4
1128  ЮТИП Технологии        4
1129           ЯКласс        4
1130            ЯрНео        4
1131    Яшин&Партнёры        4

[1132 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [65]:
# Количество работодателей (emp_cnt), у которых указана сфера деятельности (ind_name) 'Разработка программного обеспечения'
query_5_6 = f'''
select 
    i.name ind_name, 
    count(distinct ei.employer_id) emp_cnt
from 
    industries i
left join 
    public.employers_industries ei on i.id = ei.industry_id
where 
    i.name = 'Разработка программного обеспечения'
group by i.name
'''

In [66]:
# результат запроса
df_16 = pd.read_sql_query(query_5_6, connection)
display(df_16)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2721444346.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_16 = pd.read_sql_query(query_5_6, connection)


ind_name  emp_cnt
0  Разработка программного обеспечения     3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [48]:
# Код для получения списка городов-милионников
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
cities = tables[0]['Город'].values.tolist()

In [57]:
# Список регионов-миллионников (city_name) с количеством вакансий (cnt) в них
# Общее количество вакансий компании (total)
query_5_7 = f'''
select
    a.name city_name,
    count(v.id) cnt
from 
    public.employers e
left join 
    public.vacancies v on e.id = v.employer_id
left join 
    public.areas a on a.id = v.area_id
where 
    a.name in {tuple(cities)} and e.name='Яндекс'
group by 1
union all
select
    'total',
    count(v.id)
from
    public.employers e
left join 
    public.vacancies v on e.id = v.employer_id
left join 
    public.areas a on a.id = v.area_id
where 
    a.name in {tuple(cities)} and e.name='Яндекс'
group by 1
order by 2
'''

In [67]:
# результат запроса
df_17 = pd.read_sql_query(query_5_7, connection)
display(df_17)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2949694845.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_17 = pd.read_sql_query(query_5_7, connection)


city_name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

Выводы по анализу работодателей:

1. По количеству вакансий первые 5 мест занимают крупные компании: "Яндекс", "Ростелеком", "Тинькофф", "СБЕР", "Газпром нефть".

2. Наибольшее количество вакансий в крупных городах: Москва, Санкт-Петербург, Минск, Алматы, Екатеринбург.

3.  Компания "Яндекс" публикует свои вакансии в наибольшем количестве регионов и во всех городах-миллионниках.

4. Примерно 3,5 тысячи работодателей указали в качестве сферы деятельности "Разработка программного обеспечения". 
Не указана сфера деятельности примерно у 8,4 тысяч работодателей. 

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [53]:
# Количество вакансий (cnt), имеющих отношение к данным
query_6_1 = f'''
select 
    count(v.id) cnt
from 
    public.vacancies v
where 
    lower(v.name) like '%data%' 
    or lower(v.name) like '%данн%'
'''

In [68]:
# результат запроса
df_18 = pd.read_sql_query(query_6_1, connection)
display(df_18)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\37178899.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_18 = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [6]:
# Количество вакансий (ds_cnt) для дата-сайентистов
query_6_2_0 = f'''
select 
	count(v.id) ds_cnt
from 
	public.vacancies v
where
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
'''

# Количество вакансий (cnt) для дата-сайентистов уровня Junior
query_6_2_1 = f'''
select 
	count(v.id) jun_cnt
from 
	public.vacancies v
where (
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
	and (lower(v.name) like '%junior%'
	or lower(v.experience) like '%нет опыта%'
 	or lower(v.employment) like '%стажировка%')
	)
'''

In [7]:
# результат запроса
df_19 = pd.read_sql_query(query_6_2_0, connection)
display(df_19)

df_20 = pd.read_sql_query(query_6_2_1, connection)
display(df_20)

C:\Users\user\AppData\Local\Temp\ipykernel_520\1029088945.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_19 = pd.read_sql_query(query_6_2_0, connection)


ds_cnt
0     536

C:\Users\user\AppData\Local\Temp\ipykernel_520\1029088945.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_20 = pd.read_sql_query(query_6_2_1, connection)


jun_cnt
0       51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [81]:
# Количество DS-вакансий (cnt) в которых есть ключевой навык SQL или postgres
query_6_3 = f'''
select 
	count(v.id) cnt
from 
	public.vacancies v
where (
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
	and (lower(v.key_skills) like '%sql%'
	or lower(v.key_skills) like '%postgres%')
	)
'''

In [82]:
# результат запроса
df_21 = pd.read_sql_query(query_6_3, connection)
display(df_21)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\1265111920.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_21 = pd.read_sql_query(query_6_3, connection)


cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [83]:
# Количество DS-вакансий (cnt) в которых есть ключевой навык Python
query_6_4 = f'''
select 
	count(v.id) cnt
from 
	public.vacancies v
where (
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
	and lower(v.key_skills) like '%python%'
	)
'''

In [84]:
# результат запроса
df_22 = pd.read_sql_query(query_6_4, connection)
display(df_22)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\2723078428.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_22 = pd.read_sql_query(query_6_4, connection)


cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [85]:
# Количество ключевых навыков DS-вакансий (cnt).
# Удаляем Tab-разделители строк CHR(9),
# вычитаем из всей строки строку без табуляции и прибавляем число 1, 
# что равно количеству разделенных строк и вычисляем среднее значение.
query_6_5 = f'''
select 
	round(avg(length(v.key_skills) - length(replace(v.key_skills,CHR(9),''))+1),2) cnt
from 
	public.vacancies v
where 
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or v.name like '%ML%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
	and (v.key_skills is not null)
 '''

In [86]:
# результат запроса
df_23 = pd.read_sql_query(query_6_5, connection)
display(df_23)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\320684087.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_23 = pd.read_sql_query(query_6_5, connection)


cnt
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [93]:
# Средняя зарплата (salary_avg) DS-вакансий для каждого типа требуемого опыта (exp).
# Не во всех вакансиях заполнены оба поля с зарплатой, поэтому применяем
# функцию coalesce, которая возвращает первый попавшийся аргумент,
# отличный от null и вычисляем среднее значение.
query_6_6 = f'''
select 
v.experience exp,	  
round(avg(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) salary_avg
from 
	public.vacancies v
where 
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or v.name like '%ML%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
	and (v.key_skills is not null)
	and (v.salary_from is not null or v.salary_to is not null)
group by experience
order by 2 desc
'''

In [94]:
# результат запроса
df_24 = pd.read_sql_query(query_6_6, connection)
display(df_24)

C:\Users\user\AppData\Local\Temp\ipykernel_16340\3772908468.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_24 = pd.read_sql_query(query_6_6, connection)


exp  salary_avg
0       От 3 до 6 лет    243115.0
1  От 1 года до 3 лет    146704.0
2           Нет опыта     74643.0

***

Выводы по предметному анализу:

1. Количество вакансий, имеющих отношение к данным - 1771.

2. Количество вакансий для дата-сайентистов - 536, из них 51 вакансия (около 10 %) подходит для начинающих без опыта работы.

3. Примерно половина вакансий (всего 229) для дата-сайентистов, указывают в качестве ключевого навыка SQL или Postgres. 
Python в качестве ключевого навыка указывают 357 вакансий. 
В вакансиях для дата-сайентистов в среднем указывают более 6 ключевых навыков.

4. Уровень зарплаты в среднем зависит от требуемого опыта работы. 
Для дата-сайентистов без опыта работы уровень зарплаты около 75 тысяч рублей, что в 2 раза ниже, чем у специалистов с опытом от 1 года до 3 лет (около 146 тысяч рублей). У специалистов с опытом от 3 года до 6 лет уровень зарплаты около 243 тысяч рублей.


***

# Общий вывод по проекту

***

1. Для дата-сайентиста наибольшее количествоподходящих вакансий предложены в Москве и крупных городах-милионниках. 
Дата-сайентисту необходимо быть готовым к релокации в столицу или крупные города, т.к. наибольшее количество вакансий с рабочим графиком "Полный день" и типом  трудоустройства "Полная занятость".

In [8]:
# Распределение DS-вакансий по городам (топ 10)
query_7_1 = f'''
select 
a.name,
count(v.id) ds_cnt
from 
	public.vacancies v
left join public.areas a on area_id = a.id
where
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
group by 1
order by 2 desc
limit 10
'''
# результат запроса
df_25 = pd.read_sql_query(query_7_1, connection)
display(df_25)

C:\Users\user\AppData\Local\Temp\ipykernel_520\957884563.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_25 = pd.read_sql_query(query_7_1, connection)


name  ds_cnt
0           Москва     225
1  Санкт-Петербург      69
2      Новосибирск      23
3           Алматы      22
4  Нижний Новгород      20
5           Казань      19
6            Томск      11
7   Ростов-на-Дону      11
8            Минск      11
9            Пермь       8

2. Дата-сайенс требуется в различных сферах деятельности работодателей, таких как банкинг, мобильная связь, ИТ-консалтинг, соц.сети, поисковики, платежные системы, кибербезопасность итд. 
Дата-сайентист может найти работу в различных отраслях деятельности компаний.

In [9]:
# Топ 10 компаний, предлагающих DS-вакансии
query_7_2 = f'''
select 
e.name,
count(v.id) ds_cnt
from 
	public.vacancies v
left join public.employers e on v.employer_id = e.id
where
	(lower(v.name) like '%data scientist%'
	or lower(v.name) like '%data science%'
	or lower(v.name) like '%исследователь данных%'
	or lower(v.name) like '%ml%' and lower(v.name) not like '%html%'
	or lower(v.name) like '%machine learning%'
	or lower(v.name) like '%машинн%обучен%')
group by 1
order by 2 desc
limit 10
'''
# результат запроса
df_26 = pd.read_sql_query(query_7_2, connection)
display(df_26)

C:\Users\user\AppData\Local\Temp\ipykernel_520\3953659968.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_26 = pd.read_sql_query(query_7_2, connection)


name  ds_cnt
0                      СБЕР      39
1           Bell Integrator      25
2            Банк ВТБ (ПАО)      18
3                        VK      15
4     Positive Technologies      11
5                    Яндекс       9
6            EvenBet Gaming       9
7                   МегаФон       8
8                  Andersen       7
9  Иннотех, Группа компаний       6

***